engines/startrek/graphics.cpp

```diff
@@ -0,0 +1,759 @@
+	for (int i = 0; i < _numSprites; i++)
+		_sprites[i]->bitmapChanged = true;
+	drawAllSprites(updateScreen);
+}
+
+void Graphics::updateScreen() {
+	// Check if there are any pending updates to the mouse.
+	if (_mouseBitmap != _mouseBitmapLastFrame) {
+		_mouseBitmapLastFrame = _mouseBitmap;
+		_vm->_system->setMouseCursor(_mouseBitmap->pixels, _mouseBitmap->width, _mouseBitmap->height, _mouseBitmap->xoffset, _mouseBitmap->yoffset, 0);
+	}
+	if (_mouseToBeShown) {
+		CursorMan.showMouse(true);
+		_mouseToBeShown = false;
+	} else if (_mouseToBeHidden) {
+		CursorMan.showMouse(false);
+		_mouseToBeHidden = false;
+	}
+
+	if (_mouseWarpX != -1) {
+		_vm->_system->warpMouse(_mouseWarpX, _mouseWarpY);
+		_mouseWarpX = -1;
+		_mouseWarpY = -1;
+	}
+
+	_vm->_system->updateScreen();
+}
+
+Sprite *Graphics::getSpriteAt(int16 x, int16 y) {
+	for (int i = _numSprites - 1; i >= 0; i--) {
+		Sprite *sprite = _sprites[i];
+
+		if (sprite == &_lockedMouseSprite)
+			continue;
+		if (sprite->drawMode == 1) // Invisible
+			continue;
+
+		if (sprite->drawRect.contains(Common::Point(x, y))) {
+			if (sprite->drawMode == 2 || sprite->drawMode == 3) // Button or text
+				return sprite;
+
+			// For draw mode 0 only, check that we're not clicking on a transparent part.
+			int16 relX = x - sprite->drawX;
+			int16 relY = y - sprite->drawY;
+			byte pixel = sprite->bitmap->pixels[relY * sprite->bitmap->width + relX];
+			if (pixel != 0)
+				return sprite;
+		}
+	}
+
+	return nullptr;
+}
+
+void Graphics::addSprite(Sprite *sprite) {
+	if (_numSprites >= MAX_SPRITES)
+		error("addSprite: too many sprites");
+
+	// Initialize some fields
+	sprite->drawMode = 0;
+	sprite->field8 = "";
+	sprite->field16 = false;
+	sprite->bitmapChanged = true; // FIXME (delete this later?)
+
+	sprite->lastDrawRect.top = -1;
+	sprite->lastDrawRect.left = -1;
+	sprite->lastDrawRect.bottom = -2;
+	sprite->lastDrawRect.right = -2;
+
+	_sprites[_numSprites++] = sprite;
+}
+
+void Graphics::delSprite(Sprite *sprite) {
+	for (int i = 0; i < _numSprites; i++) {
+		if (sprite != _sprites[i])
+			continue;
+
+		_numSprites--;
+		_sprites[i] = _sprites[_numSprites];
+		return;
+	}
+
+	error("delSprite: sprite not in list");
+}
+
+void Graphics::pushSprites() {
+	if (_pushedNumSprites != -1)
+		error("Tried to push sprites more than once");
+	_pushedNumSprites = _numSprites;
+	memcpy(_pushedSprites, _sprites, sizeof(_sprites));
+
+	_numSprites = 0;
+}
+
+void Graphics::popSprites() {
+	if (_pushedNumSprites == -1)
+		error("Tried to pop sprites without a prior push");
+	_numSprites = _pushedNumSprites;
+	memcpy(_sprites, _pushedSprites, sizeof(_sprites));
+
+	_pushedNumSprites = -1;
+}
+
+byte *Graphics::getFontGfx(char c) {
+	return _font->getCharData(c & 0xff);
+}
+
+
+void Graphics::copyBackgroundScreen() {
+	drawDirectToScreen(_backgroundImage);
+}
+
+void Graphics::drawDirectToScreen(SharedPtr<Bitmap> bitmap) {
+	int xoffset = bitmap->xoffset;
```

Comment:
What is the point of these variables, when you're doing other references to bitmap variables in the same call below?